File: go/singleton.go
package main

import "sync"

// 싱글톤 예시
type Logger struct {
	data string
}

var instance *Logger
var once sync.Once

func GetInstance() *Logger {
	once.Do(func() {
		instance = &Logger{}
	})
	return instance
}

func (l *Logger) SetData(data string) {
	l.data = data
}

func main() {
	logger1 := GetInstance()
	logger1.SetData("Hello")
	logger2 := GetInstance()
	logger2.SetData("World")

	println(logger1.data) // World
	println(logger2.data) // World
}

File: js/singleton.ts
class Logger {
  private static instance: Logger;
  private data: string;

  private constructor() {
    this.data = "";
  }

  public static getInstance(): Logger {
    if (!Logger.instance) {
      this.instance = new Logger();
    }
    return Logger.instance;
  }

  public setData(data: string): void {
    this.data = data;
  }

  public getData(): string {
    return this.data;
  }
}

const l1 = Logger.getInstance();
const L2 = Logger.getInstance();

l1.setData("Hello");
console.log(l1.getData()); // Hello
console.log(L2.getData()); // Hello
